File: hackerrank/medium/0007-encryption/python/solution.py
## solution.py

from math import floor, ceil, sqrt


def encryption(s: str) -> str:
    work = s.replace(" ", "")
    L = len(work)

    root = sqrt(L)
    rows = floor(root)
    cols = ceil(root)

    if rows * cols < L:
        rows += 1

    grid = [work[i : i + cols] for i in range(0, L, cols)]
    columns = []

    for c in range(cols):
        col_chars = []
        for r in range(len(grid)):
            if c < len(grid[r]):
                col_chars.append(grid[r][c])
        columns.append("".join(col_chars))

    return " ".join(columns)


# --- demo run ---
if __name__ == "__main__":
    print(encryption("haveaniceday"))  # hae and via ecy
    print(encryption("feedthedog"))    # fto ehg ee dd
    print(encryption("chillout"))      # clu hlt io

File: hackerrank/medium/0007-encryption/js/solution.mjs
// solution.mjs

export function encryption(s) {
  const work = s.replace(/ /g, "");
  const L = work.length;

  const root = Math.sqrt(L);
  let rows = Math.floor(root);
  let cols = Math.ceil(root);

  if (rows * cols < L) {
    rows += 1;
  }

  // Build grid rows
  const grid = [];
  for (let i = 0; i < L; i += cols) {
    grid.push(work.slice(i, i + cols));
  }

  const resultCols = [];

  for (let c = 0; c < cols; c++) {
    let colStr = "";
    for (let r = 0; r < grid.length; r++) {
      if (c < grid[r].length) {
        colStr += grid[r][c];
      }
    }
    resultCols.push(colStr);
  }

  return resultCols.join(" ");
}

// --- demo run ---
if (import.meta.url === `file://${process.argv[1]}`) {
  console.log(encryption("haveaniceday")); // hae and via ecy
  console.log(encryption("feedthedog"));   // fto ehg ee dd
  console.log(encryption("chillout"));     // clu hlt io
}
